ta/gatekeeper/gatekeeper_ta.c

```diff
@@ -0,0 +1,767 @@
+	memset(params, 0, sizeof(params));
+
+	res = TEE_OpenTASession(&uuid, TEE_TIMEOUT_INFINITE,
+				param_types, params, &sess,
+				&return_origin);
+	if (res) {
+		EMSG("Failed to connect to keymaster");
+		goto exit;
+	}
+
+	param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+				      TEE_PARAM_TYPE_MEMREF_OUTPUT,
+				      TEE_PARAM_TYPE_NONE,
+				      TEE_PARAM_TYPE_NONE);
+	memset(&params, 0, sizeof(params));
+
+	params[0].memref.buffer = dummy;
+	params[0].memref.size = sizeof(dummy);
+
+	params[1].memref.buffer = auth_token_key_data;
+	params[1].memref.size = sizeof(auth_token_key_data);
+
+	res = TEE_InvokeTACommand(sess, TEE_TIMEOUT_INFINITE,
+				  KM_GET_AUTHTOKEN_KEY,
+				  param_types, params, &return_origin);
+	if (res) {
+		EMSG("Failed in keymaster");
+		goto close_sess;
+	}
+
+	if (params[1].memref.size != sizeof(auth_token_key_data)) {
+		EMSG("Wrong auth_token key size");
+		res = TEE_ERROR_CORRUPT_OBJECT;
+
+		goto close_sess;
+	}
+
+	TEE_InitRefAttribute(&attr, TEE_ATTR_SECRET_VALUE,
+			     auth_token_key_data, sizeof(auth_token_key_data));
+
+	res = TEE_PopulateTransientObject(key, &attr, 1);
+	if (res)
+		EMSG("Failed to set auth_token key attributes");
+
+close_sess:
+	TEE_CloseTASession(sess);
+exit:
+	return res;
+}
+
+static void TA_MintAuthToken(struct hw_auth_token *auth_token,
+			     int64_t timestamp,
+			     secure_id_t user_id,
+			     secure_id_t authenticator_id,
+			     uint64_t challenge)
+{
+	TEE_Result res;
+
+	struct hw_auth_token token;
+	TEE_ObjectHandle auth_token_key = TEE_HANDLE_NULL;
+
+	const uint8_t *to_sign = (const uint8_t *)&token;
+	const uint32_t to_sign_sz = sizeof(token) - sizeof(token.hmac);
+
+	token.version = HW_AUTH_TOKEN_VERSION;
+	token.challenge = challenge;
+	token.user_id = user_id;
+	token.authenticator_id = authenticator_id;
+	token.authenticator_type = TEE_U32_TO_BIG_ENDIAN((uint32_t)
+							 HW_AUTH_PASSWORD);
+	token.timestamp =  TEE_U64_TO_BIG_ENDIAN(timestamp);
+	memset(token.hmac, 0, sizeof(token.hmac));
+
+	res = TEE_AllocateTransientObject(TEE_TYPE_HMAC_SHA256,
+					  TEE_SHA256_HASH_SIZE * 8,
+					  &auth_token_key);
+	if (res) {
+		EMSG("Failed to allocate auth_token key");
+		goto exit;
+	}
+
+	res = TA_GetAuthTokenKey(auth_token_key);
+	if (res) {
+		EMSG("Failed to get auth_token key from keymaster");
+		goto free_key;
+	}
+
+	res = TA_ComputeSignature(token.hmac, sizeof(token.hmac),
+				  auth_token_key, to_sign, to_sign_sz);
+	if (res) {
+		EMSG("Failed to compute auth_token signature");
+		memset(token.hmac, 0, sizeof(token.hmac));
+		goto free_key;
+	}
+
+free_key:
+	TEE_FreeTransientObject(auth_token_key);
+exit:
+	memcpy(auth_token, &token, sizeof(token));
+}
+
+static TEE_Result TA_DoVerify(const struct password_handle *expected_handle,
+			      const uint8_t *password,
+			      uint32_t password_length)
+{
+	TEE_Result res;
+	struct password_handle password_handle;
+
+	if (!password_length) {
+		res = TEE_ERROR_BAD_PARAMETERS;
```

Comment:
Since we've had security issues with non-initialized "res", I'd recommend to initialize `res` at 372 with this value instead.

Reply:
fixed